akka-stream-tests/src/test/scala/akka/stream/scaladsl/One2OneBidiFlowSpec.scala

```diff
@@ -0,0 +1,84 @@
+/**
+ * Copyright (C) 2015 Typesafe Inc. <http://www.typesafe.com>
+ */
+package akka.stream.scaladsl
+
+import scala.concurrent.Await
+import scala.concurrent.duration._
+import org.scalactic.ConversionCheckedTripleEquals
+import akka.stream.ActorMaterializer
+import akka.stream.testkit.AkkaSpec
+import akka.stream.testkit._
+
+class One2OneBidiFlowSpec extends AkkaSpec with ConversionCheckedTripleEquals {
+  import FlowGraph.Implicits._
+
+  implicit val mat = ActorMaterializer()
+
+  /**
+   * Wraps a flow with one-to-one guarding, which
+   *
+   * 1. triggers an `OutputTruncationException` if the wrapped stream terminates
+   *    before exactly one output has been produced for each input element
+   * 2. Backpressures surplus output elements (i.e. ones that are being produced without
+   *    having a corresponding input element)
+   * 3. makes sure that output elements are being produced in lockstep with input elements,
+   *    i.e. no output element is ever being produced before the corresponding input element hasn't been received
+   *
+   * // TODO: make this available on Flows; Where do we place it best?
+   */
+  def one2one[I, O, Mat](flow: Flow[I, O, Mat]): Flow[I, O, Mat] =
+    Flow(flow) { implicit b ⇒
+      f ⇒
+        val o2o = b.add(One2OneBidiFlow[I, O])
+        o2o.out1 ~> f ~> o2o.in2
+        (o2o.in1, o2o.out2)
+    }
+
+  "A One2OneBidiFlow" must {
+
+    def test(flow: Flow[Int, Int, Unit]) =
+      Source(List(1, 2, 3)).via(flow).grouped(10).runWith(Sink.head)
+
+    "be fully transparent for valid one-to-one streams" in {
+      val f = one2one(Flow[Int].map(_ * 2))
+      Await.result(test(f), 1.second) should ===(Seq(2, 4, 6))
+    }
+
+    "be fully transparent to errors" in {
+      val f = one2one(Flow[Int].map(x ⇒ 10 / (x - 2)))
+      an[ArithmeticException] should be thrownBy Await.result(test(f), 1.second)
+    }
+
+    "trigger an `OutputTruncationException` if the wrapped stream terminates early" in {
+      val f = one2one(Flow[Int].filter(_ < 3))
+      a[One2OneBidiFlow.OutputTruncationException.type] should be thrownBy Await.result(test(f), 1.second)
+    }
+
+    "backpressure surplus output elements" in {
+      val f = one2one(Flow[Int].mapConcat(x ⇒ x :: x :: Nil))
+      Await.result(test(f), 1.second) should ===(Seq(1, 1, 2))
```

Comment:
It backpressures and then truncates, which amounts to the surplus elements being dropped, yes.